### PR TITLE
Fast path interleaved in conversion

### PR DESCRIPTION
**Pull request by SkySkimmer**

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
@coqbot ci minimize

**Comment by SkySkimmer** [comment]
fiat crypto legacy error further reduces to
~~~coq
Axiom T:Prop.

Lemma foo
  {x y : T -> Type}
  : True.
Proof.
  match goal with
  | [ H : ?T, H' : ?T |- _ ] => idtac
  end.
~~~

**Comment by SkySkimmer** [comment]
@coqbot ci minimize

**Comment by SkySkimmer** [comment]
Gonna need an equations expert to look at that failure

**Comment by ppedrot** [comment]
It's a good thing you have the unique inhabitant of that type next to you in your office.

**Comment by mattam82** [comment]
https://github.com/mattam82/Coq-Equations/pull/524

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
@mattam82 that didn't work https://gitlab.com/coq/coq/-/jobs/3551713857#L1467

**Comment by SkySkimmer** [comment]
@coqbot ci minimize ci-category_theory

**Comment by SkySkimmer** [comment]
@coqbot ci minimize

**Comment by SkySkimmer** [comment]
So what do we do with this? Is equations impossible to fix?

**Comment by SkySkimmer** [comment]
Actually this bug seems not really about equations. Conversion is called with an evar map which has some univ variable `v := ...` and a term which mentions `v`, but `v` is not in the evar map's ugraph.
The cause is in 2 steps:
- univ minimization removes the newly defined universes from the local univ set 
- regenerating the ustate (ie update_sigma_univs) relies on the local univ set 

so we get a non-equations example:
~~~coq
Lemma foo : True.
Proof.
  pose (forall (x:Type(*@u*)) (z:(fun y:Type@{_}(*@v*) => tt) x = tt), True).
  (* {u v} |= u <= v, undefined v *)

  assert True by abstract exact I.
  (* {u} |= [], v := u
     v is still in the ustate's ugraph *)

  Monomorphic Axiom bla : Type.
  (* ustate regenerated from global env, v is not in the ustate's ugraph! *)

  Check (eq_refl : P = (forall (x:Type) (z:(fun y:Type => tt) x = tt), True)).
  (* anomaly in univ.repr, universe undefined *)
~~~

Possible fixes:
- make the conversion fast path use the universe substitution in some way (add a field to the evar handler I guess)
- make minimization keep the defined variables in the local set (restrict will remove them so most of the time behaviour should be identical I guess)
- change update_sigma_univs to also declare variables from the substitution (pushing this mostly because it needs the least code change and to see if other bugs reveal themselves)

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
>Possible fixes:

putting needs:progress so consider the other options

**Comment by SkySkimmer** [comment]
>make minimization keep the defined variables in the local set (restrict will remove them so most of the time behaviour should be identical I guess)

This is probably the easier to reason about choice when thinking about ustate invariants. Probably just needs to remove the `remove u ctx` at https://github.com/ppedrot/coq/blob/bb4f304848e04c492d98db5da0bdb1895cecc191/engine/univMinim.ml#L265

>make the conversion fast path use the universe substitution in some way (add a field to the evar handler I guess)

Probably least impactful choice, IMO it is reasonably principled as having to pass the kernel some extra info when we call it with `Unsafe.to_constr` is ok.

>change update_sigma_univs to also declare variables from the substitution (pushing this mostly because it needs the least code change and to see if other bugs reveal themselves)

This is just a hack ;)

**Comment by SkySkimmer** [comment]
Changes I notice since the original bench:
- bedrock2 was failing and is now 15% faster
- fiat-crypto-with-bedrock was failing and is now 2% faster (looks real from CPU instruction change)
- hott was failing and is now 1% faster (looks real from CPU instruction change)
- category-theory was failing and is now 2% slower (looks real from CPU instruction change)
- stdlib was 2% faster and is now 3.8% slower (not sure if real either time)
- coqutil was failing and is now noise range changed
- metacoq was passing with noise range changes and is now failing because the benched commit included https://github.com/coq/coq/pull/17177

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
>make minimization keep the defined variables in the local set

I looked a bit more at this but it's harder than it looks.

Let's try just making it so the fast path doesn't care about the ugraph and worry about this later.

**Comment by SkySkimmer** [comment]
The first vst bench in a long time

Apparently using the ugraph in the fast path is useless so this is ready to merge

File: kernel/reduction.ml [hunk 1]
```diff
@@ -353,9 +353,82 @@ let irr_flex env = function
...
+    Array.equal_norefl (fun c1 c2 -> compare_under CONV e1 c1 e2 c2) t1 t2
+    && compare_under CONV e1 def1 e2 def2
+    && compare_under CONV e1 ty1 e2 ty2
+  | (Rel _ | Meta _ | Var _ | Sort _ | Prod _ | Lambda _ | LetIn _ | App _
+    | Proj _ | Evar _ | Const _ | Ind _ | Construct _ | Case _ | Fix _
+    | CoFix _ | Int _ | Float _| Array _), _ -> false
+
+
+let rec fast_test infos pb lft1 term1 lft2 term2 = match fterm_of term1, fterm_of term2 with
+  (* is it el_shft or el_liftn? *)
```
Comment (silene) [inline kernel/reduction.ml:418]: What about this comment? (I do think `el_shft` is correct here, but I would not bet on it.)

Comment (ppedrot) [reply kernel/reduction.ml:418]: This is the correct one.

File: kernel/reduction.ml [hunk 2]
```diff
@@ -353,9 +353,89 @@ let irr_flex env = function
...
+        (subst_instance_sort (snd e1) s1)
+        (subst_instance_sort (snd e2) s2)
+    | CUMUL ->
+      UGraph.check_leq_sort univs
+        (subst_instance_sort (snd e1) s1)
+        (subst_instance_sort (snd e2) s2)
+    end
+  | Prod (_,t1,c1), Prod (_,t2,c2) ->
+    compare_under univs CONV e1 t1 e2 t2
+    && compare_under univs CUMUL (usubs_lift e1) c1 (usubs_lift e2) c2
```
Comment (ppedrot) [inline kernel/reduction.ml]: This `CUMUL` should read `pb` here.

Comment (ppedrot) [reply kernel/reduction.ml]: OK, so please remove it so that we only care about trivial equality comparisons.

**Comment by ppedrot** [comment]
@SkySkimmer I am fiddling with this PR to see whether we can have a slightly more general criterion than asking the lifts to be equal. Basically, I use this variant:
```ocaml
let rec fast_test infos pb lft1 term1 lft2 term2 = match fterm_of term1, fterm_of term2 with
  (* is it el_shft or el_liftn? *)
  | FLIFT (i, term1), (FLIFT _ | FCLOS _) -> fast_test infos pb (el_shft i lft1) term1 lft2 term2
  | FCLOS _, FLIFT (j, term2) -> fast_test infos pb lft1 term1 (el_shft j lft2) term2
  | FCLOS (c1, (e1,u1)), FCLOS (c2, (e2,u2)) ->
    let e1 = Esubst.lift_subst (fun _ _ -> ()) lft1 e1 in
    let e2 = Esubst.lift_subst (fun _ _ -> ()) lft2 e2 in
    compare_under pb (e1, u1) c1 (e2, u2) c2
  | _ -> false
end
```
The weird thing is that this results in assertion failures because the conversion engine is breaking the invariant that FCLOS terms live in the right lifted environment. That is, if there are subcalls to conversions such that we have on one side `FCLOS(t, σ)` with `Γ ⊢ t : A`, `σ : Δ ⊢ Γ` against some lift `lft` but it does not hold that `lft : Ξ ⊢ Δ` for some Ξ. This looks like a big invariant violation to me. I'm not sure we can weaponize this into a proof of unsoundness in master but maybe your patch is enough for that.

A small reproducible example:
```coq
Axiom admit : False.

Definition iff (A : Prop) := inhabited (A -> True).

Unset Kernel Term Sharing.

Definition foo :
eq
  (forall (_ : True) (x : unit), inhabited (True -> True))
  ((fun A : Prop => True -> A) (forall _ : unit, iff True))
:= eq_refl.
```
The variable `A` gets substituted with the term `FCLOS(iff true, σ)` where the lift applied in conversion disagree with the substitution σ.

Am I expecting an incorrect invariant, or is there something very wrong going on?

**Comment by SkySkimmer** [comment]
I have no clue what the lift stuff is doing tbh

**Comment by SkySkimmer** [comment]
@coqbot run full co

**Comment by SkySkimmer** [comment]
@coqbot run full ci

File: kernel/reduction.ml (hunk 2 above)
Comment (SkySkimmer) [reply kernel/reduction.ml]: changed

**Review by ppedrot — APPROVED** [review]
*À dieu vat.*

**Comment by ppedrot** [comment]
@coqbot merge now

**Pull request merged**